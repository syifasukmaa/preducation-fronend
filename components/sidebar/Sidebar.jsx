'use client'
import React, { useState } from 'react'
import Image from 'next/image'
import { RxHamburgerMenu } from 'react-icons/rx'
import { IoMdClose } from 'react-icons/io'
import { LuLayoutDashboard, LuLogOut } from 'react-icons/lu'
import { SiGoogleclassroom } from 'react-icons/si'
import { usePathname, useRouter } from 'next/navigation'
import { signOut } from 'next-auth/react'
import Swal from 'sweetalert2'


export default function Sidebar() {
  const url = usePathname()
  const router = useRouter()

  const [open, setOpen] = useState(false)

  const toggleSidebar = () => {
    setOpen(!open)
  }

  const sidebarItems = [
    { id: '/admin/dashboard', icon: <LuLayoutDashboard />, label: 'Dashboard' },
    {
      id: '/admin/course',
      icon: <SiGoogleclassroom />,
      label: 'Kelola Kelas',
    },
  ]

  const handleClick = (itemId) => {
    router.push(itemId)
  }

  const handleLogOut = () => {
    Swal.fire({
      title: 'Apakah yakin?',
      icon: 'warning',
      showCancelButton: true,
      confirmButtonColor: '#167F71',
      cancelButtonColor: '#d33',
      confirmButtonText: 'Keluar',
      cancelButtonText: 'Batal',
      customClass: {
        title: 'custom-title-class',
        content: 'custom-text-class',
        popup: 'custom-card-class',
        cancelButton: 'custom-button-cancel',
        confirmButton: 'custom-button-confirm',
      },
    }).then((result) => {
      if (result.isConfirmed) {
        signOut()
      }
    })
  }

  return (
    <aside
      className={`bg-primary-dark-blue text-white w-60 md:w-3/12 lg:w-[15%]
      transition-all duration-300 ease-in-out  ${open ? 'translate-x-0' : '-translate-x-3/4 md:translate-x-0'}
      overflow-y-auto z-50 shadow-lg fixed top-0 left-0 bottom-0 md:pt-0 `}
    >
      <button
        onClick={toggleSidebar}
        title="hamburger-button"
        className="fixed z-50 px-3 py-2 text-white rounded md:hidden right-4 top-14 bg-orange-05"
      >
        {open ? <IoMdClose /> : <RxHamburgerMenu />}
      </button>

      <Image
        src="/img/iconPreducation.png"
        // className={`overflow-hidden transition-all w-[65%] ml-5 md:m-auto ${open ? 'w-32' : 'w-0'}`}
        className={`md:m-auto`}
        alt="Icon Belajar"
        width={110}
        height={110}
        priority
      />

      <nav className="py-2 mt-6 md:mt-3">
        <ul className={`${open ? '' : 'flex flex-col justify-end items-end pr-3 md:pr-0 md:block'}`}>
          {sidebarItems?.map((item, index) => (
            <li
              key={index}
              className={` text-white py-3 ${url.startsWith(item.id) ? 'bg-orange-05' : ''}`}

            >
              <button
                href={item.id}
                className={`flex items-center px-4 ${url.startsWith(item.id) ? 'ml-4' : ''}`}
                onClick={() => handleClick(item.id)}
              data-testid={item.label}
                
              >
                <span>{item.icon}</span>
                <span className={`ml-3 ${open ? 'block' : 'hidden md:block'} font-bold`}>{item.label}</span>
              </button>
            </li>
          ))}

          <li className={`text-white py-3 bg-orange-05} active:bg-orange-05`}>
            <button
              className={`flex items-center px-4`}
              onClick={handleLogOut}
              data-testid="handle-logout"
            >
              <span>
                <LuLogOut />
              </span>
              <span className={`ml-3 ${open ? 'block' : 'hidden md:block'} font-bold`}>Keluar</span>
            </button>
          </li>
        </ul>
      </nav>
    </aside>
  )
}
