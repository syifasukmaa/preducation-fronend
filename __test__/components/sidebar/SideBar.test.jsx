import Sidebar from '@/components/sidebar/Sidebar'
import { fireEvent, render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { signOut } from 'next-auth/react'
import { usePathname } from 'next/navigation'
import Swal from 'sweetalert2'

jest.mock('next/navigation', () => ({
  ...jest.requireActual('next/navigation'),
  useRouter: jest.fn(() => ({
    push: jest.fn(),
  })),
  usePathname: jest.fn(),
}))

jest.mock('next-auth/react', () => ({
  signOut: jest.fn(),
}))

jest.mock('sweetalert2', () => ({
  fire: jest.fn(),
}))

describe('Navbar Component', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('Should renders the sidebar component', () => {
    usePathname.mockReturnValue('/admin/dashboard')

    const container = render(<Sidebar />)
    expect(container).toMatchSnapshot()
  })

  it('toggles the Sidebar when the hamburger button is clicked', async () => {
    const { getByTitle } = render(<Sidebar />)
    const hamburgerButton = getByTitle('hamburger-button')
    expect(hamburgerButton).toBeInTheDocument()
    expect(hamburgerButton.innerHTML).toContain(
      '<svg stroke="currentColor" fill="none" stroke-width="0" viewBox="0 0 15 15" height="1em" width="1em" xmlns="http://www.w3.org/2000/svg"><path fill-rule="evenodd" clip-rule="evenodd" d="M1.5 3C1.22386 3 1 3.22386 1 3.5C1 3.77614 1.22386 4 1.5 4H13.5C13.7761 4 14 3.77614 14 3.5C14 3.22386 13.7761 3 13.5 3H1.5ZM1 7.5C1 7.22386 1.22386 7 1.5 7H13.5C13.7761 7 14 7.22386 14 7.5C14 7.77614 13.7761 8 13.5 8H1.5C1.22386 8 1 7.77614 1 7.5ZM1 11.5C1 11.2239 1.22386 11 1.5 11H13.5C13.7761 11 14 11.2239 14 11.5C14 11.7761 13.7761 12 13.5 12H1.5C1.22386 12 1 11.7761 1 11.5Z" fill="currentColor"></path></svg>'
    )
    await userEvent.click(hamburgerButton)

    expect(hamburgerButton.innerHTML).toContain(
      '<svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 512 512" height="1em" width="1em" xmlns="http://www.w3.org/2000/svg"><path d="M405 136.798L375.202 107 256 226.202 136.798 107 107 136.798 226.202 256 107 375.202 136.798 405 256 285.798 375.202 405 405 375.202 285.798 256z"></path></svg>'
    )
  })

  it('handles navigation when a side bar item is clicked', async () => {
    usePathname.mockReturnValue('/admin/dashboard')
    const { getByTestId } = render(<Sidebar />)

    //expect after clicked, will redirect to page /admin/course
    fireEvent.click(getByTestId('Dashboard'))
    expect(screen.getByText(/Kelola Kelas/i)).toBeInTheDocument()
  })

  it('Handles logout when "Keluar" is clicked ', async () => {
    await Swal.fire.mockResolvedValueOnce({ isConfirmed: true })

    const { getByTestId } = render(<Sidebar />)
    const logoutButton = getByTestId('handle-logout')

    fireEvent.click(logoutButton)

    expect(Swal.fire).toHaveBeenCalledWith({
      title: 'Apakah yakin?',
      icon: 'warning',
      showCancelButton: true,
      confirmButtonColor: '#167F71',
      cancelButtonColor: '#d33',
      confirmButtonText: 'Keluar',
      cancelButtonText: 'Batal',
    })

    await Swal.fire.mockResolvedValueOnce({ isConfirmed: true })

    expect(signOut).toHaveBeenCalled()
  })
})
