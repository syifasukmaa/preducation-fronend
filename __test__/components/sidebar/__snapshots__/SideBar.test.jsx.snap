// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar Component Should renders the sidebar component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <aside
        class="bg-primary-dark-blue text-white w-60 md:w-3/12 lg:w-[15%]
      transition-all duration-300 ease-in-out  -translate-x-3/4 md:translate-x-0
      overflow-y-auto z-50 shadow-lg fixed top-0 left-0 bottom-0 md:pt-0 "
      >
        <button
          class="fixed z-50 px-3 py-2 text-white rounded md:hidden right-4 top-14 bg-orange-05"
          title="hamburger-button"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 15 15"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M1.5 3C1.22386 3 1 3.22386 1 3.5C1 3.77614 1.22386 4 1.5 4H13.5C13.7761 4 14 3.77614 14 3.5C14 3.22386 13.7761 3 13.5 3H1.5ZM1 7.5C1 7.22386 1.22386 7 1.5 7H13.5C13.7761 7 14 7.22386 14 7.5C14 7.77614 13.7761 8 13.5 8H1.5C1.22386 8 1 7.77614 1 7.5ZM1 11.5C1 11.2239 1.22386 11 1.5 11H13.5C13.7761 11 14 11.2239 14 11.5C14 11.7761 13.7761 12 13.5 12H1.5C1.22386 12 1 11.7761 1 11.5Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </button>
        <img
          alt="Icon Belajar"
          class="md:m-auto"
          data-nimg="1"
          decoding="async"
          fetchpriority="high"
          height="110"
          src="/_next/image?url=%2Fimg%2FiconPreducation.png&w=256&q=75"
          srcset="/_next/image?url=%2Fimg%2FiconPreducation.png&w=128&q=75 1x, /_next/image?url=%2Fimg%2FiconPreducation.png&w=256&q=75 2x"
          style="color: transparent;"
          width="110"
        />
        <nav
          class="py-2 mt-6 md:mt-3"
        >
          <ul
            class="flex flex-col justify-end items-end pr-3 md:pr-0 md:block"
          >
            <li
              class=" text-white py-3 bg-orange-05"
              data-testid="Dashboard"
            >
              <button
                class="flex items-center px-4"
                href="/admin/dashboard"
              >
                <span>
                  <svg
                    fill="none"
                    height="1em"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <rect
                      height="9"
                      rx="1"
                      width="7"
                      x="3"
                      y="3"
                    />
                    <rect
                      height="5"
                      rx="1"
                      width="7"
                      x="14"
                      y="3"
                    />
                    <rect
                      height="9"
                      rx="1"
                      width="7"
                      x="14"
                      y="12"
                    />
                    <rect
                      height="5"
                      rx="1"
                      width="7"
                      x="3"
                      y="16"
                    />
                  </svg>
                </span>
                <span
                  class="ml-3 hidden md:block font-bold"
                >
                  Dashboard
                </span>
              </button>
            </li>
            <li
              class=" text-white py-3 "
              data-testid="Kelola Kelas"
            >
              <button
                class="flex items-center px-4"
                href="/admin/course"
              >
                <span>
                  <svg
                    fill="currentColor"
                    height="1em"
                    role="img"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <title />
                    <path
                      d="M1.6367 1.6367C.7322 1.6367 0 2.369 0 3.2734v17.4532c0 .9045.7322 1.6367 1.6367 1.6367h20.7266c.9045 0 1.6367-.7322 1.6367-1.6367V3.2734c0-.9045-.7322-1.6367-1.6367-1.6367H1.6367zm.545 2.1817h19.6367v16.3632h-2.7266v-1.0898h-4.9102v1.0898h-12V3.8184zM12 8.1816c-.9046 0-1.6367.7322-1.6367 1.6368 0 .9045.7321 1.6367 1.6367 1.6367.9046 0 1.6367-.7322 1.6367-1.6367 0-.9046-.7321-1.6368-1.6367-1.6368zm-4.3633 1.9102c-.6773 0-1.2285.5493-1.2285 1.2266 0 .6772.5512 1.2265 1.2285 1.2265.6773 0 1.2266-.5493 1.2266-1.2265 0-.6773-.5493-1.2266-1.2266-1.2266zm8.7266 0c-.6773 0-1.2266.5493-1.2266 1.2266 0 .6772.5493 1.2265 1.2266 1.2265.6773 0 1.2285-.5493 1.2285-1.2265 0-.6773-.5512-1.2266-1.2285-1.2266zM12 12.5449c-1.179 0-2.4128.4012-3.1484 1.0059-.384-.1198-.8043-.1875-1.2149-.1875-1.3136 0-2.7285.695-2.7285 1.5586v.8965h14.1836v-.8965c0-.8637-1.4149-1.5586-2.7285-1.5586-.4106 0-.831.0677-1.2149.1875-.7356-.6047-1.9694-1.0059-3.1484-1.0059Z"
                    />
                  </svg>
                </span>
                <span
                  class="ml-3 hidden md:block font-bold"
                >
                  Kelola Kelas
                </span>
              </button>
            </li>
            <li
              class="text-white py-3 bg-orange-05} active:bg-orange-05 flex items-center px-4"
              data-testid="handle-logout"
            >
              <span>
                <svg
                  fill="none"
                  height="1em"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M9 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h4"
                  />
                  <polyline
                    points="16 17 21 12 16 7"
                  />
                  <line
                    x1="21"
                    x2="9"
                    y1="12"
                    y2="12"
                  />
                </svg>
              </span>
              <span
                class="ml-3 hidden md:block font-bold"
              >
                Keluar
              </span>
            </li>
          </ul>
        </nav>
      </aside>
    </div>
  </body>,
  "container": <div>
    <aside
      class="bg-primary-dark-blue text-white w-60 md:w-3/12 lg:w-[15%]
      transition-all duration-300 ease-in-out  -translate-x-3/4 md:translate-x-0
      overflow-y-auto z-50 shadow-lg fixed top-0 left-0 bottom-0 md:pt-0 "
    >
      <button
        class="fixed z-50 px-3 py-2 text-white rounded md:hidden right-4 top-14 bg-orange-05"
        title="hamburger-button"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 15 15"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M1.5 3C1.22386 3 1 3.22386 1 3.5C1 3.77614 1.22386 4 1.5 4H13.5C13.7761 4 14 3.77614 14 3.5C14 3.22386 13.7761 3 13.5 3H1.5ZM1 7.5C1 7.22386 1.22386 7 1.5 7H13.5C13.7761 7 14 7.22386 14 7.5C14 7.77614 13.7761 8 13.5 8H1.5C1.22386 8 1 7.77614 1 7.5ZM1 11.5C1 11.2239 1.22386 11 1.5 11H13.5C13.7761 11 14 11.2239 14 11.5C14 11.7761 13.7761 12 13.5 12H1.5C1.22386 12 1 11.7761 1 11.5Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <img
        alt="Icon Belajar"
        class="md:m-auto"
        data-nimg="1"
        decoding="async"
        fetchpriority="high"
        height="110"
        src="/_next/image?url=%2Fimg%2FiconPreducation.png&w=256&q=75"
        srcset="/_next/image?url=%2Fimg%2FiconPreducation.png&w=128&q=75 1x, /_next/image?url=%2Fimg%2FiconPreducation.png&w=256&q=75 2x"
        style="color: transparent;"
        width="110"
      />
      <nav
        class="py-2 mt-6 md:mt-3"
      >
        <ul
          class="flex flex-col justify-end items-end pr-3 md:pr-0 md:block"
        >
          <li
            class=" text-white py-3 bg-orange-05"
            data-testid="Dashboard"
          >
            <button
              class="flex items-center px-4"
              href="/admin/dashboard"
            >
              <span>
                <svg
                  fill="none"
                  height="1em"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <rect
                    height="9"
                    rx="1"
                    width="7"
                    x="3"
                    y="3"
                  />
                  <rect
                    height="5"
                    rx="1"
                    width="7"
                    x="14"
                    y="3"
                  />
                  <rect
                    height="9"
                    rx="1"
                    width="7"
                    x="14"
                    y="12"
                  />
                  <rect
                    height="5"
                    rx="1"
                    width="7"
                    x="3"
                    y="16"
                  />
                </svg>
              </span>
              <span
                class="ml-3 hidden md:block font-bold"
              >
                Dashboard
              </span>
            </button>
          </li>
          <li
            class=" text-white py-3 "
            data-testid="Kelola Kelas"
          >
            <button
              class="flex items-center px-4"
              href="/admin/course"
            >
              <span>
                <svg
                  fill="currentColor"
                  height="1em"
                  role="img"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <title />
                  <path
                    d="M1.6367 1.6367C.7322 1.6367 0 2.369 0 3.2734v17.4532c0 .9045.7322 1.6367 1.6367 1.6367h20.7266c.9045 0 1.6367-.7322 1.6367-1.6367V3.2734c0-.9045-.7322-1.6367-1.6367-1.6367H1.6367zm.545 2.1817h19.6367v16.3632h-2.7266v-1.0898h-4.9102v1.0898h-12V3.8184zM12 8.1816c-.9046 0-1.6367.7322-1.6367 1.6368 0 .9045.7321 1.6367 1.6367 1.6367.9046 0 1.6367-.7322 1.6367-1.6367 0-.9046-.7321-1.6368-1.6367-1.6368zm-4.3633 1.9102c-.6773 0-1.2285.5493-1.2285 1.2266 0 .6772.5512 1.2265 1.2285 1.2265.6773 0 1.2266-.5493 1.2266-1.2265 0-.6773-.5493-1.2266-1.2266-1.2266zm8.7266 0c-.6773 0-1.2266.5493-1.2266 1.2266 0 .6772.5493 1.2265 1.2266 1.2265.6773 0 1.2285-.5493 1.2285-1.2265 0-.6773-.5512-1.2266-1.2285-1.2266zM12 12.5449c-1.179 0-2.4128.4012-3.1484 1.0059-.384-.1198-.8043-.1875-1.2149-.1875-1.3136 0-2.7285.695-2.7285 1.5586v.8965h14.1836v-.8965c0-.8637-1.4149-1.5586-2.7285-1.5586-.4106 0-.831.0677-1.2149.1875-.7356-.6047-1.9694-1.0059-3.1484-1.0059Z"
                  />
                </svg>
              </span>
              <span
                class="ml-3 hidden md:block font-bold"
              >
                Kelola Kelas
              </span>
            </button>
          </li>
          <li
            class="text-white py-3 bg-orange-05} active:bg-orange-05 flex items-center px-4"
            data-testid="handle-logout"
          >
            <span>
              <svg
                fill="none"
                height="1em"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M9 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h4"
                />
                <polyline
                  points="16 17 21 12 16 7"
                />
                <line
                  x1="21"
                  x2="9"
                  y1="12"
                  y2="12"
                />
              </svg>
            </span>
            <span
              class="ml-3 hidden md:block font-bold"
            >
              Keluar
            </span>
          </li>
        </ul>
      </nav>
    </aside>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
